compiler/rustc_typeck/src/collect.rs

```diff
@@ -2469,43 +2469,187 @@ fn compute_sig_of_foreign_fn_decl<'tcx>(
-            if ty.is_simd() {
-                let snip = tcx
-                    .sess
-                    .source_map()
-                    .span_to_snippet(ast_ty.span)
-                    .map_or_else(|_| String::new(), |s| format!(" `{}`", s));
-                tcx.sess
-                    .struct_span_err(
-                        ast_ty.span,
-                        &format!(
-                            "use of SIMD type{} in FFI is highly experimental and \
-                             may result in invalid code",
-                            snip
-                        ),
-                    )
-                    .help("add `#![feature(simd_ffi)]` to the crate attributes to enable")
-                    .emit();
-            }
-        };
+    // Using SIMD types in FFI signatures requires the signature
+    // to have appropriate `#[target_feature]`'s enabled.
+    if abi != abi::Abi::RustIntrinsic && abi != abi::Abi::PlatformIntrinsic {
         for (input, ty) in iter::zip(decl.inputs, fty.inputs().skip_binder()) {
-            check(&input, ty)
+            simd_ffi_check(tcx, def_id, &input, ty)
         }
         if let hir::FnRetTy::Return(ref ty) = decl.output {
-            check(&ty, fty.output().skip_binder())
+            simd_ffi_check(tcx, def_id, &ty, fty.output().skip_binder())
         }
     }
 
     fty
 }
 
+/// Returns `Ok()` if the target-feature allows using the SIMD type on C FFI.
+/// Otherwise, returns `Err(Some())` if the target_feature needs to be enabled or
+/// or `Err(None)` if it's unsupported.
+fn simd_ffi_feature_check(
+    target: &str,
+    simd_width: u64,
+    simd_elem_width: u64,
+    feature: String,
+) -> Result<(), Option<&'static str>> {
+    match target {
+        t if t.contains("x86") => {
+            // FIXME: this needs to be architecture dependent and
+            // should probably belong somewhere else:
+            // * on mips: 16 => msa,
+            // * wasm: 16 => simd128
+            match simd_width {
+                8 if feature.contains("mmx")
+                    || feature.contains("sse")
+                    || feature.contains("ssse")
+                    || feature.contains("avx") =>
+                {
+                    Ok(())
+                }
+                8 => Err(Some("mmx")),
+                16 if feature.contains("sse")
+                    || feature.contains("ssse")
+                    || feature.contains("avx") =>
+                {
+                    Ok(())
+                }
+                16 => Err(Some("sse")),
+                32 if feature.contains("avx") => Ok(()),
+                32 => Err(Some("avx")),
+                64 if feature.contains("avx512") => Ok(()),
+                64 => Err(Some("acx512")),
+                _ => Err(None),
+            }
+        }
+        t if t.contains("arm") => {
+            match simd_width {
+                // 32-bit arm does not support vectors with 64-bit wide elements
```

Comment:
32-bit ARM supports `i64` vectors but not `f64` vectors. LLVM will still pass `f64x2` parameters in NEON registers though so it's fine to just always allow it.